design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,159 @@
+# KLIP-43: TIMESTAMP Data Type Support
+
+**Author**: @jzaralim | 
+**Release Target**: 0.15 | 
+**Status**: In Discussion | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/6649
+
+**tl;dr:** _KSQL currently converts time data into long values. This KLIP introduces a TIMESTAMP
+data type to handle time data instead._
+
+## Motivation and background
+
+With Connect integration, KSQL processes data from various sources and sinks. Most other databases
+have a TIMESTAMP or DATETIME type, which KSQL currently converts into long values. This sometimes
+causes problems when sinking into a database. Supporting TIMESTAMP types would make moving time data
+between KSQL and other data sources/sinks smoother and less error-prone.
+
+## What is in scope
+### Required
+
+* Add TIMESTAMP type to KSQL
+* Serialization and de-serialization of TIMESTAMPs to Avro, JSON and Delimited formats
+* Built-in functions to support TIMESTAMP usage
+
+### Future enhancements
+
+* Give window units (HOUR, DAY etc) a TIMESTAMP value
+* Add DATE and TIME types
+* UDFs such as DAY and HOUR
+
+## What is not in scope
+Changing the ROWTIME data type. We will eventually want this to happen, but that is a separate discussion.
+
+## Public APIS
+
+The TIMESTAMP data type will store a time without timezone information. It will be useable in any
+place STRING or INTEGER could be used:
+```roomsql
+CREATE STREAM stream_name (col1 TIMESTAMP, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field TIMESTAMP>) AS ...
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the Java Date type within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
```

Comment:
Can't this be an implicit cast? Like:
```
insert into (ts) values ('2020-11-20 14:34:00');
```

That above doesn't require a UDF to convert to timestamps. 

Or is the function used for string columns instead of literals? For this, other SQL systems use `TO_TIMESTAMP`, which converts a String to a Timestamp.

Reply:
The reason for this instead of a cast is to support different string formats. Perhaps having a date format property in KSQL could make supporting TIMESTAMP/STRING casting possible?

Reply:
we could copy what mySQL does here: https://dev.mysql.com/doc/refman/8.0/en/datetime.html but it would require implementing a local time zone to do the conversion as well

Reply:
Yeah supporting string/timestamp casting should work. I think we should keep these functions though, in case a user wants to convert strings in a format that isn't recognized by standard formatters. .